### PR TITLE
feat: reorder character selection screen

### PR DESCRIPTION
**Pull request by r-o-b-o-t-o**
## Changes Proposed:
-  Add an `order` column to the `characters` table, allowing players to reorder their characters in the character selection screen

## Issues Addressed:

None

## Tests Performed:

- Applied SQL and C++ patches
- Tested on Windows 10 client and server

## How to Test the Changes:

1. Create multiple characters in an account, eg:
image
2. Change their `order` column, for example:
```sql
UPDATE `characters` SET `order` = 1 WHERE `name` = "One";
UPDATE `characters` SET `order` = 2 WHERE `name` = "Two";
UPDATE `characters` SET `order` = 3 WHERE `name` = "Three";
UPDATE `characters` SET `order` = 4 WHERE `name` = "Four";
```
3. Log out and log back in
4. The characters should be reordered according to the `order` column:
image
5. (optional) If you want to reset back to the default order, use
```sql
UPDATE `characters` SET `order` = NULL WHERE `name` IN ("One", "Two", "Three", "Four");
```

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: data/sql/updates/pending_db_characters/rev_1624468542752528700.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+INSERT INTO `version_db_characters` (`sql_rev`) VALUES ('1624468542752528700');
+
+ALTER TABLE `characters` DROP IF EXISTS `order`;
+ALTER TABLE `characters` ADD `order` TINYINT(1) NULL DEFAULT NULL AFTER `grantableLevels`;
```
Comment (Helias) [inline data/sql/updates/pending_db_characters/rev_1624468542752528700.sql:4]: you can remove the (1) of TINYINT, it will be deprecated soon

**Comment by Helias** [comment]
We could create a way to handle this order via acore-cms @Yehonal @Dalanth, ofc it's a low-priority thing.
If someone made a command to change the order of his character it will be easier to make an interface in acore-cms to change this order.

File: data/sql/updates/pending_db_characters/rev_1624468542752528700.sql (hunk 1 above)
Comment (Kitzunu) [reply data/sql/updates/pending_db_characters/rev_1624468542752528700.sql:4]: not "Can" -> "Should" :P

**Comment by r-o-b-o-t-o** [comment]
> We could create a way to handle this order via acore-cms @Yehonal @nicoaravena , ofc it's a low-priority thing.
> If someone made a command to change the order of his character it will be easier to make an interface in acore-cms to change this order.

Don't worry I've already done it 😄
I'll create a PR in acore-cms if this one gets merged!

File: data/sql/updates/pending_db_characters/rev_1624468542752528700.sql [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+            (TABLE_NAME = @tablename)
+            AND (TABLE_SCHEMA = @dbname)
+            AND (COLUMN_NAME = @columnname)
+    ) > 0,
+    "SELECT 1 WHERE false",
+    CONCAT("ALTER TABLE `", @tablename, "` ADD `", @columnname, "` TINYINT NULL DEFAULT NULL AFTER `grantableLevels`;")
+));
+PREPARE alterIfNotExists FROM @preparedStatement;
+EXECUTE alterIfNotExists;
+DEALLOCATE PREPARE alterIfNotExists;
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_characters/rev_1624468542752528700.sql:19]: What does this do over

```sql
ALTER TABLE characters
  ADD order TINYINY NULL DEFAULT NULL AFTER granableLevels

Comment (r-o-b-o-t-o) [reply data/sql/updates/pending_db_characters/rev_1624468542752528700.sql:19]: It ensures we can run it multiple times without throwing an error, it's the column version of "always do a DELETE query before an INSERT"

Comment (r-o-b-o-t-o) [reply data/sql/updates/pending_db_characters/rev_1624468542752528700.sql:19]: As you can see in the commit history, at first I did
```sql
ALTER TABLE `characters` DROP IF EXISTS `order`;
```
before the `ADD`, and it worked well, but the `IF EXISTS` part only works in mariadb, this doesn't exist in mysql, so I had to resort to this

Comment (Kitzunu) [reply data/sql/updates/pending_db_characters/rev_1624468542752528700.sql:19]: It does exist, or is it different? https://dev.mysql.com/doc/refman/8.0/en/drop-table.html

Comment (Kitzunu) [reply data/sql/updates/pending_db_characters/rev_1624468542752528700.sql:19]: Maybe it is only for tables and not columns

Comment (Kitzunu) [reply data/sql/updates/pending_db_characters/rev_1624468542752528700.sql:19]: nvm. It's only for DROP and not ALTER https://bugs.mysql.com/bug.php?id=10789

Comment (r-o-b-o-t-o) [reply data/sql/updates/pending_db_characters/rev_1624468542752528700.sql:19]: Yes, this is only for tables. There's a [feature request for columns from 2005](https://bugs.mysql.com/bug.php?id=10789), still nothing 😒

Comment (r-o-b-o-t-o) [reply data/sql/updates/pending_db_characters/rev_1624468542752528700.sql:19]: Oh you linked the same thing as I was typing it haha

**Comment by Kitzunu** [comment]
I always find it weird how just adding a table can just magically determine this lol

**Comment by Sedatyf** [comment]
I tested it as @r-o-b-o-t-o described
- First I created those characters
image

- Then I applied this request
```sql
UPDATE `characters` SET `order` = 3 WHERE `name` = "One";
UPDATE `characters` SET `order` = 1 WHERE `name` = "Two";
UPDATE `characters` SET `order` = 2 WHERE `name` = "Three";
```

And all characters have been sorted like mentioned in the SQL request ✔️ 
image

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by FrancescoBorzi** [comment]
excellent job as usual @r-o-b-o-t-o and @Sedatyf

**Comment by necropola** [comment]
@r-o-b-o-t-o Nice! It's now live on CC. Sadly the wiki still isn't updated.
https://www.azerothcore.org/wiki/characters

**Comment by r-o-b-o-t-o** [comment]
> @r-o-b-o-t-o Nice! It's now live on CC. Sadly the wiki still isn't updated.
> https://www.azerothcore.org/wiki/characters

Done, thanks for reminding me 😅
https://github.com/azerothcore/wiki/pull/594

**Comment by Tony931023** [comment]
It is already in the list of updates of the character table, PR https://github.com/azerothcore/wiki/pull/584 , thanks to @r-o-b-o-t-o I took the description of the order column.

**Pull request merged**